[Flax] Add FlaxMBart

stancld opened the pull request:
# What does this PR do?
This PR adds flax implementation of MBart.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@patrickvonplaten  @patil-suraj

stancld committed:
Copy BART to MBart and rename some stuff

stancld committed:
Add copy statements pointing to FlaxBart

stancld committed:
Update/add some common files

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+        return outputs
+
+
+class FlaxMBartEncoderLayerCollection(nn.Module):
```
Can we add copied from here?

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+        return outputs
+
+
+class FlaxMBartDecoderLayerCollection(nn.Module):
```
Can we add copied from here?

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+            FlaxMBartDecoderLayer(self.config, name=str(i), dtype=self.dtype)
+            for i in range(self.config.decoder_layers)
+        ]
+        self.layerdrop = self.config.encoder_layerdrop
```
```suggestion
        self.layerdrop = self.config.decoder_layerdrop
```

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+
+    def setup(self):
+        self.dropout_layer = nn.Dropout(rate=self.config.dropout)
+        self.layerdrop = self.config.encoder_layerdrop
```
think we can remove this. We don't need layerdrop here.

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+
+    def setup(self):
+        self.dropout_layer = nn.Dropout(rate=self.config.dropout)
+        self.layerdrop = self.config.decoder_layerdrop
```
same comment as above.

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1732 @@
+        )
+
+
+class FlaxMBartModule(nn.Module):
```
Think we can add copied from here as well.

patil-suraj commented on tests/test_modeling_flax_mbart.py:
```diff
@@ -0,0 +1,417 @@
+
+
+@require_flax
+class BartHeadTests(unittest.TestCase):
```
```suggestion
class MBartHeadTests(unittest.TestCase):
```

patil-suraj commented on tests/test_modeling_flax_mbart.py:
```diff
@@ -0,0 +1,417 @@
+        self.assertEqual(outputs["logits"].shape, expected_shape)
+
+    def test_lm_uneven_forward(self):
+        config = BartConfig(
```
```suggestion
        config = MBartConfig(
```

patil-suraj commented on tests/test_modeling_flax_mbart.py:
```diff
@@ -0,0 +1,417 @@
+        )
+
+        batch_size = input_ids.shape[0]
+        config = BartConfig(
```
```suggestion
        config = MBartConfig(
```

patil-suraj reviewed (commented):
Thanks a lot for adding this @stancld!

It's looking great overall, I left a few comments. Specifically

- The order of layer norm and attention and
- Could we add as much copied from statements as possible?

stancld committed:
Update shift_tokens_rigth + fix imports

* Fix shift_tokens_right method according to MBart implementation

* Update shift_tokens_right in tests accordingly

* Fix the import issue and update docs file
* make style quality

stancld committed:
Merge master into flax_mbart

stancld committed:
Do some minor changes according to patil-suraj suggestions

* Change the order of normalization layer and attention

* Add some copu statementes

stancld commented:
@patil-suraj Thank you a lot for your suggestions. I fixed the order of attention and normalization layers and some other minor bugs. Also added some additional copy statements.

I also changed the `shift_tokens_right` method as this one looks to be different for the MBart models as they don't have a single `decoder_start_token_id` in contrast to other Bart-like models. => This change, however, currently leads to some issues within the `generate` method.

stancld committed:
Update generate method and add integration test for mBart

patil-suraj commented on src/transformers/generation_flax_utils.py:
```diff
@@ -238,14 +238,14 @@ def generate(
             model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(input_ids, model_kwargs)
             # prepare decoder_input_ids for generation
-            input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+            if decoder_start_token_id is not None:
+                input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+            else:
+                input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4")
```
`decoder_start_token_id` should never be `None`. For mBART it is the target language id, which could be passed an argument to the `generate`. So let's remove this.

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1744 @@
+            )
+            hidden_states = self.dropout_layer(hidden_states, deterministic=deterministic)
+            hidden_states = residual + hidden_states
+            hidden_states = self.encoder_attn_layer_norm(hidden_states)
```
This should also be applied before `encoder_attn`

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1744 @@
+        >>> inputs = tokenizer([ARTICLE_TO_SUMMARIZE], max_length=1024, return_tensors='jax')
+
+        >>> # Generate Summary
+        >>> summary_ids = model.generate(inputs['input_ids']).sequences
```
```suggestion
        >>> summary_ids = model.generate(inputs['input_ids'], decoder_start_token_id=tokenizer.lang_code_to_id[tgt_lang]).sequences
```

patil-suraj reviewed (approved):
Looks good to me! Just left a couple of comments that need to be taken care of before merging.

I will run all slow tests and push the checkpoints to the hub before merging.

stancld committed:
Make a few updates after a review

Besides, add `lang_code_to_id` to MBartTokenizeFast

stancld committed:
Merge master into flax_mbart and resolve conflicts

stancld committed:
fix-copies; make style quality

stancld committed:
Merge master into flax_mbart

stancld committed:
Merge branch 'master' into flax_mbart

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1741 @@
+            dtype=self.dtype,
+        )
+        self.layers = FlaxMBartEncoderLayerCollection(self.config, self.dtype)
+        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
```
```suggestion
        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
        self.layer_norm = nn.LayerNorm(dtype=self.dtype)        
```

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1741 @@
+            return_dict=return_dict,
+        )
+
+        if not return_dict:
+            return outputs
+
+        return FlaxBaseModelOutput(
+            last_hidden_state=outputs.last_hidden_state,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
```
```suggestion
        last_hidden_states = outputs[0]
        last_hidden_states = self.layer_norm(last_hidden_states)

        if not return_dict:
            return (last_hidden_states,) + outputs[1:]

        return FlaxBaseModelOutput(
            last_hidden_state=last_hidden_states,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1741 @@
+        )
+
+        self.layers = FlaxMBartDecoderLayerCollection(self.config, self.dtype)
+        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
```
```suggestion
        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
        self.layer_norm = nn.LayerNorm(dtype=self.dtype)
```

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1741 @@
+            return_dict=return_dict,
+        )
+
+        if not return_dict:
+            return outputs
+
+        return FlaxBaseModelOutputWithPastAndCrossAttentions(
+            last_hidden_state=outputs.last_hidden_state,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            cross_attentions=outputs.cross_attentions,
+        )
```
```suggestion
        last_hidden_states = outputs[0]
        last_hidden_states = self.layer_norm(last_hidden_states)

        if not return_dict:
            return (last_hidden_states,) + outputs[1:]

        return FlaxBaseModelOutput(
            last_hidden_state=last_hidden_states,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

patil-suraj commented on tests/test_modeling_flax_mbart.py:
```diff
@@ -0,0 +1,463 @@
+            attention_mask=model_inputs.attention_mask,
+            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
+            num_beams=2,
+        )
```
```suggestion
        ).sequences
```

patil-suraj reviewed (commented):
The encoder and decoder reqiores an extra layer_norm at the end

patil-suraj committed:
Apply suggestions from code review

patil-suraj commented on tests/test_modeling_flax_mbart.py:
```diff
@@ -0,0 +1,463 @@
+        generated_ids = self.model.generate(
+            model_inputs.input_ids,
+            attention_mask=model_inputs.attention_mask,
+            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
```
```suggestion
            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
            early_stopping=True,
```

patil-suraj committed:
Apply suggestions from code review

patil-suraj commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1748 @@
+        return FlaxBaseModelOutput(
+            last_hidden_state=last_hidden_states,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
```
```suggestion
            attentions=outputs.attentions,
            cross_attentions=outputs.cross_attentions,
```

patil-suraj committed:
Apply suggestions from code review

patil-suraj committed:
fix output type, style

patil-suraj commented:
@stancld I pushed a couple of commits to add the `layer_norm` in encoder and decoder. Now, all slow tests are passing.
@patrickvonplaten  could you please take a final look?

patrickvonplaten commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1747 @@
+        [prev_output_tokens[i, eos_idx] for i, eos_idx in enumerate(index_of_eos)]
+    ).squeeze()
+    # for loop basically does jax-compatible version of prev_output_tokens[:, 1:] = prev_output_tokens[:, :-1].clone()
+    for i in range(prev_output_tokens.shape[1], 0, -1):
```
nice!

patrickvonplaten commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1747 @@
+        return hidden_states.reshape(hidden_states.shape[:2] + (self.embed_dim,))
+
+    @nn.compact
+    def _concatenate_to_cache(self, key, value, query, attention_mask):
```
can we use some `# Copied from` statements here?

patrickvonplaten commented on src/transformers/models/mbart/modeling_flax_mbart.py:
```diff
@@ -0,0 +1,1747 @@
+            decoder_position_ids,
+        )["params"]
+
+    def init_cache(self, batch_size, max_length, encoder_outputs):
```
Add copied from statement?

patrickvonplaten reviewed (approved):
Very nice!

patil-suraj committed:
add copied from

patil-suraj committed:
Merge branch 'master' into flax_mbart

patil-suraj committed:
resolve conflicts
